File: Heeseop An/2023-01-20-3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()

        for i, n in enumerate(nums):
            if i > 0 and n == nums[i - 1]:
                continue

            l = i + 1
            r = len(nums) - 1

            while l < r:
                total = n + nums[l] + nums[r]
                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    result.append([n, nums[l], nums[r]])
                    l += 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
        return result

File: Dave Min/2023-01-20-3Sum.cs
public class Solution
{
    public IList<IList<int>> ThreeSum(int[] nums)
    {
        Array.Sort(nums);
        HashSet<int> hset = new HashSet<int>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (!hset.Contains(nums[i])) hset.Add(nums[i]);
        }

        HashSet<string> duplChecker = new HashSet<string>();
        IList<IList<int>> resultList = new List<IList<int>>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (nums[i] > 0) break;

            for (int j = i + 1; j < nums.Length; j++)
            {
                if (0 - (nums[j] + nums[i]) < nums[j]) continue;
                if (hset.Contains(0 - (nums[j] + nums[i])))
                {
                    var result = new List<int>() { nums[i], nums[j], 0 - (nums[j] + nums[i]) };
                    if (!duplChecker.Contains(string.Concat(result)))
                    {
                        duplChecker.Add(string.Concat(result));
                        resultList.Add(result);
                    }
                }
            }
        }
        return resultList;
    }
}